mempool/cat/spec.md

```diff
@@ -0,0 +1,101 @@
+
+### Purpose
+
+The objective of such a protocol is to transport transactions from the author (usually a client) to a proposed block, optimizing both latency and throughput i.e. how quickly can a transaction be proposed (and committed) and how many transactions can be transported into a block at once.
+
+Typically the mempool serves to receive inbound transactions via an RPC endpoint, gossip them to all nodes in the network (regardless of whether they are capable of proposing a block or not), and stage groups of transactions to both consensus and the application to be included in a block.
+
+### Assumptions
+
+The following are assumptions inherited from existing Tendermint mempool protocols:
+
+- `CheckTx` should be seen as a simple gatekeeper to what transactions enter the pool to be gossiped and staged. It is non-deterministic: one node may reject a transaction that another node keeps.
+- Applications implementing `CheckTx` are responsible for replay protection (i.e. the same transaction being present in multiple blocks). The mempool ensures that within the same block, no duplicate transactions can exist.
+- The underlying p2p layer guarantees eventually reliable broadcast. A transaction need only be sent once to eventually reach the target peer.
+
+### Messages
+
+The CAT protocol extends on the existing mempool implementations by introducing two new protobuf messages:
+
+```protobuf
+message SeenTx {
+  bytes tx_key = 1;
+  optional string from = 2;
+}
+
+message WantTx {
+  bytes tx_key = 1;
+}
+```
+
+Both `SeenTx` and `WantTx` contain the sha256 hash of the raw transaction bytes. `SeenTx` also contains an optional `p2p.ID` that corresponds to the peer that the node recieved the tx from. The only validation for both is that the byte slice of the `tx_key` MUST have a length of 32.
+
+Both messages are sent across a new channel with the ID: `byte(0x31)`. This enables cross compatability as discussed in greater detail below.
+
+> **Note:**
+> The term `SeenTx` is used over the more common `HasTx` because the transaction pool contains sophisticated eviction logic. TTL's, higher priority transactions and reCheckTx may mean that a transaction pool *had* a transaction but does not have it any more. Semantically it's more appropriate to use `SeenTx` to imply not the presence of a transaction but that the node has seen it and dealt with it accordingly.
+
+### Outbound logic
+
+A node in the protocol has two distinct modes: "broadcast" and "request/response". When a node receives a transaction via RPC (or specifically through `CheckTx`), it assumed that it is the only recipient from that client and thus will immediately send that transaction, after validation, to all connected peers. Afterwards, only "request/response" is used to disseminate that transaction to everyone else.
```

Comment:
[Question]
If  a transaction is sent to a validator, and that validator is connected to 50 peers. Then, that transaction will only make it to the 50 connected peers. If we assume a validator set of 150, 100 validators will not see it until it is included in a block. Won't this delay the amount of time users will need to wait before their transaction gets included?

Reply:
> If a transaction is sent to a validator, and that validator is connected to 50 peers. Then, that transaction will only make it to the 50 connected peers.

Only initially. Then each of those 50 connected peers will broadcast `SeenTx` messages to the peers they are connected with which means all nodes should eventually see these `SeenTx` messages. For the nodes that don't have the transaction. They can send a `WantTx` message, essentially requesting it from that peer. That way all nodes will eventually receive all transactions.

Reply:
Thanks a lot :+1: